### PR TITLE
Support for Vapor 4

### PR DESCRIPTION
**Pull request by PoshAlpaca**
Adds support for Vapor 4. Originally implemented by rabc. I've combined it with our custom `filterGeographyDistanceWithin` and `sortByDistance` functions.

### CODE REVIEW

**Comment by PoshAlpaca** [comment]
@0xTim  ready for review 🎉

File: Package.swift [hunk 1]
```diff
@@ -1,30 +1,31 @@
-// swift-tools-version:4.1
+// swift-tools-version:5.1
```
Comment (0xTim) [inline Package.swift:1]: Let's just set this to 5.5 to make life easier

File: Package.swift [hunk 2]
```diff
@@ -1,30 +1,31 @@
...
 import PackageDescription
 
 let package = Package(
-    name: "FluentPostGIS",
+    name: "fluent-postgis",
+    platforms: [
+       .macOS(.v10_15)
```
Comment (0xTim) [inline Package.swift:7]: Let's do v12 to allow us to add async/await stuff easily if we want to

File: README.md [hunk 3]
```diff
@@ -1,38 +1,58 @@
...
-PostGIS support for [FluentPostgreSQL](https://github.com/vapor/fluent-postgresql) and [Vapor](https://github.com/vapor/vapor)
+PostGIS support for [fluent-postgres-driver](https://github.com/vapor/fluent-postgres-driver) and [Vapor 4](https://github.com/vapor/vapor)
 
 # Installation
 
 ## Swift Package Manager
 
 ```swift
-.package(url: "https://github.com/plarson/fluent-postgis.git", .branch("master"))
+.package(url: "https://github.com/rabc/fluent-postgis.git", .branch("vapor_4"))
```
Comment (0xTim) [inline README.md:13]: Let's set this for the Broken Hands fork set up

File: README.md [hunk 4]
```diff
@@ -1,38 +1,58 @@
 # FluentPostGIS
 
-[![Build Status](https://travis-ci.org/plarson/fluent-postgis.svg?branch=master)](https://travis-ci.org/plarson/fluent-postgis)
 ![Platforms](https://img.shields.io/badge/platforms-Linux%20%7C%20OS%20X-blue.svg)
 ![Package Managers](https://img.shields.io/badge/package%20managers-SwiftPM-yellow.svg)
-[![Twitter dizm](https://img.shields.io/badge/twitter-dizm-green.svg)](http://twitter.com/dizm)
 
-PostGIS support for [FluentPostgreSQL](https://github.com/vapor/fluent-postgresql) and [Vapor](https://github.com/vapor/vapor)
+PostGIS support for [fluent-postgres-driver](https://github.com/vapor/fluent-postgres-driver) and [Vapor 4](https://github.com/vapor/vapor)
```
Comment (0xTim) [inline README.md:6]: Let's add a sentence about it being a fork of the PostGIS package that provides geographic support on top of the geometric support from the upstream repo

File: README.md [hunk 5]
```diff
@@ -1,38 +1,58 @@
...
+
 # Models
-Add ```GISGeographicPoint2D``` to your models
+Add a type to your model
 ```swift
-final class User: PostgreSQLModel {
+final class UserLocation: Model {
+    static let schema = "user_location"
+    
+    @ID(custom: "id", generatedBy: .database)
```
Comment (0xTim) [inline README.md:33]: We might as well make this a UUID as that's the default in Fluent

File: README.md [hunk 6]
```diff
@@ -1,38 +1,58 @@
...
+}
+```
+
+Then use its data type in the `Migration`:
+
+```swift
+struct UserLocationMigration: Migration {
+    func prepare(on database: Database) -> EventLoopFuture<Void> {
+        return database.schema(UserLocation.schema)
+            .field("id", .int, .identifier(auto: true))
```
Comment (0xTim) [inline README.md:46]: Again UUID like the model

File: README.md [hunk 7]
```diff
@@ -44,10 +64,10 @@ final class User: PostgreSQLModel {
 |GeometricGeometryCollection2D|GeographicGeometryCollection2D|
 
 # Queries
-Query locations using ```ST_DWithin```
+Query using any of the filter functions:
 ```swift        
-let searchLocation = GISGeographicPoint2D(longitude: -71.060316, latitude: 48.432044)
-try User.query(on: conn).filterGeometryDistanceWithin(\User.location, searchLocation, 1000).all().wait()
+let searchLocation = GeometricPoint2D(x: 1, y: 2)
+try UserLocation.query(on: conn).filterGeometryDistanceWithin(\.$location, user.location, 1000).all().wait()
```
Comment (0xTim) [inline README.md:70]: We define a search location but don't actually use it

File: README.md [hunk 8]
```diff
@@ -44,10 +64,10 @@ final class User: PostgreSQLModel {
 |GeometricGeometryCollection2D|GeographicGeometryCollection2D|
 
 # Queries
-Query locations using ```ST_DWithin```
+Query using any of the filter functions:
 ```swift        
-let searchLocation = GISGeographicPoint2D(longitude: -71.060316, latitude: 48.432044)
-try User.query(on: conn).filterGeometryDistanceWithin(\User.location, searchLocation, 1000).all().wait()
+let searchLocation = GeometricPoint2D(x: 1, y: 2)
```
Comment (0xTim) [inline README.md:69]: Let's use a better lat/lon for this, something like Buckingham Palace etc

File: Sources/FluentPostGIS/Queries/QueryBuilder+Contains.swift [hunk 9]
```diff
@@ -18,88 +13,50 @@ extension QueryBuilder where
...
-    ///
-    /// - parameters:
-    ///     - value: Value to filter by.
-    ///     - field: Name to a field on the model to filter.
-    /// - returns: Query builder for chaining.
-    @discardableResult
-    private func filterGeometryContains(_ value: Database.QueryFilterValue, _ field: Database.QueryField) -> Self {
-        return self.filter(custom: Database.queryGeometryContains(value, field))
-    }
+       ///
```
Comment (0xTim) [inline Sources/FluentPostGIS/Queries/QueryBuilder+Contains.swift:24]: Indentation seems off here in the diff

File: Sources/FluentPostGIS/Support/EnablePostGISMigration.swift [hunk 10]
```diff
@@ -1,14 +1,27 @@
+import FluentKit
 import Foundation
-import FluentPostgreSQL
+import SQLKit
 
 public struct EnablePostGISMigration: Migration {
-    public typealias Database = PostgreSQLDatabase
+    public init() {}
```
Comment (0xTim) [inline Sources/FluentPostGIS/Support/EnablePostGISMigration.swift:6]: Just double check the name of the migration is computed to be the same in Vapor 3 and Vapor 4 to make sure we don't try to run this again

File: Tests/FluentPostGISTests/GeometryTests.swift [hunk 11]
```diff
@@ -1,116 +1,62 @@
...
+        defer { try! UserLocationMigration().revert(on: conn).wait() }
+
         let point = GeometricPoint2D(x: 1, y: 2)
 
-        var user = UserLocation(id: nil, location: point)
-        user = try user.save(on: conn).wait()
-        
+        let user = UserLocation()
+        user.location = point
+        _ = try user.save(on: conn).wait()
```
Comment (0xTim) [inline Tests/FluentPostGISTests/GeometryTests.swift:13]: If async/await is available to us here we should use it

File: Tests/FluentPostGISTests/GeometryTests.swift [hunk 12]
```diff
@@ -1,116 +1,62 @@
...
+
         let fetched = try UserLocation.find(1, on: conn).wait()
         XCTAssertEqual(fetched?.location, point)
 
-        let all = try UserLocation.query(on: conn).filterGeometryDistanceWithin(\UserLocation.location, user.location, 1000).all().wait()
+        let all = try UserLocation.query(on: conn).filterGeometryDistanceWithin(\.$location, user.location, 1000).all().wait()
         XCTAssertEqual(all.count, 1)
```
Comment (0xTim) [inline Tests/FluentPostGISTests/GeometryTests.swift:51]: a/a if possible

File: Tests/FluentPostGISTests/GeometryTests.swift [hunk 13]
```diff
@@ -1,116 +1,62 @@
...
         let point2 = GeometricPoint2D(x: 2, y: 3)
         let point3 = GeometricPoint2D(x: 3, y: 2)
         let lineString = GeometricLineString2D(points: [point, point2, point3, point])
         let polygon = GeometricPolygon2D(exteriorRing: lineString, interiorRings: [lineString, lineString])
 
-        var user = UserArea(id: nil, area: polygon)
-        user = try user.save(on: conn).wait()
+        let user = UserArea()
+        user.area = polygon
+        _ = try user.save(on: conn).wait()
```
Comment (0xTim) [inline Tests/FluentPostGISTests/GeometryTests.swift:51]: `save()` returns a `Void` so the `_` can be removed

File: Tests/FluentPostGISTests/XCTestManifests.swift [hunk 14]
```diff
@@ -2,7 +2,7 @@ import XCTest
 
 extension GeometryTests {
     static let __allTests = [
-//        ("testGeometryCollection", testGeometryCollection),
+        ("testGeometryCollection", testGeometryCollection),
```
Comment (0xTim) [inline Tests/FluentPostGISTests/XCTestManifests.swift:5]: Let's delete this file and rely on test discovery instead (assuming it works with the subclassing!)

**Review by 0xTim — CHANGES_REQUESTED** [review]
Great work! Few comments and things that can be updated. Some other things to look at are:
* use async/await throughout the tests assuming it's available to us. Updating the manifest to 5.5 and macOS 12 will help
* we don't have any CI any more, we should set up GH actions to run the tests to make sure things work

**Comment by PoshAlpaca** [comment]
@0xTim Thanks for the feedback! Can you take a look again?

**Comment by PoshAlpaca** [comment]
Also, I was thinking to extend `DatabaseSchema.DataType` with static properties so that fields in migrations can be declared like so:

```swift
.field("location", .geometricPoint2D)
```

instead of:

```swift
.field("location", GeometricPoint2D.dataType)
```

What do you think?

File: .github/workflows/ci.yml [hunk 15]
```diff
@@ -0,0 +1,19 @@
+name: CI
+on:
+- push
```
Comment (0xTim) [inline .github/workflows/ci.yml:3]: We should add pull request here as well

File: .github/workflows/ci.yml [hunk 16]
```diff
@@ -0,0 +1,19 @@
...
+        image: postgis/postgis
+        ports:
+          - "5432:5432"
+        env:
+          POSTGRES_USER: fluentpostgis
+          POSTGRES_PASSWORD: fluentpostgis
+          POSTGRES_DB: postgis_tests
+    steps:
+    - uses: actions/checkout@v3
+    - run: swift test --enable-test-discovery
```
Comment (0xTim) [inline .github/workflows/ci.yml:19]: Don't need test discovery that's default since 5.4

**Comment by 0xTim** [comment]
@PoshAlpaca yep I think that's a great idea to provide a cleaner API

**Review by 0xTim — CHANGES_REQUESTED** [review]
Looking great! Last couple of changes just to CI then we're good. We should have CI passing in this PR

And let's add the data type stuff as well if it's simple

**Review by 0xTim — APPROVED** [review]
Looks great! Might need to tweak the `on` section of CI to get it to run or wait until it's merged. Happy to merge however and fix in another PR if necessary

**Pull request merged**